ykrt/src/compile/jitc_yk/gdb.rs

```diff
@@ -0,0 +1,275 @@
+#[repr(C)]
+pub struct JitDescriptor {
+    version: u32,
+    action_flag: u32,
+    relevant_entry: *mut JitCodeEntry,
+    first_entry: *mut JitCodeEntry,
+}
+
+/// An instance of [JitDescriptor] with a special symbol name that gdb recognises specially.
+///
+/// Only [JitDescriptorAccessor] should directly access this.
+#[allow(non_upper_case_globals)]
+#[no_mangle]
+pub static mut __jit_debug_descriptor: JitDescriptor = JitDescriptor {
+    version: 1,
+    action_flag: 0,
+    relevant_entry: ptr::null_mut(),
+    first_entry: ptr::null_mut(),
+};
+
+/// Due to the way the gdb JIT API works, we are unable to have Rust handle the synchronisation of
+/// `__jit_debug_descriptor` for us.
+unsafe impl Send for JitDescriptor {}
+unsafe impl Sync for JitDescriptor {}
+unsafe impl Send for JitCodeEntry {}
+unsafe impl Sync for JitCodeEntry {}
+
+/// Wraps `__jit_debug_descriptor`, ensuring that accesses are synchronised.
+struct JitDescriptorAccessor {
+    mtx: Mutex<()>,
+}
+
+impl JitDescriptorAccessor {
+    fn with_mut<F>(&self, mut f: F)
+    where
+        F: FnMut(*mut JitDescriptor),
+    {
+        let lock = self.mtx.lock().unwrap();
+        unsafe { f(ptr::addr_of_mut!(__jit_debug_descriptor)) };
+        drop(lock);
+    }
+}
+
+static JIT_DESCRIPTOR_ACCESSOR: JitDescriptorAccessor = JitDescriptorAccessor {
+    mtx: Mutex::new(()),
+};
+
+/// The JITted code registration hook.
+///
+/// GDB regognises calls to this function to detect when JITted code is being loaded.
+#[inline(never)]
+#[no_mangle]
+pub extern "C" fn __jit_debug_register_code() {}
+
+/// Describes the mapping from a line in the source file to a virtual address.
+#[derive(Debug)]
+#[deku_derive(DekuWrite)]
+struct LineInfo {
+    vaddr: usize,
+    line_num: c_int,
+}
+
+/// Our custom gdb "symbol file".
+///
+/// Instances of this get serialised for gdb to read. Note that gdb runs in a separate address
+/// space, so we can't put pointers from this address space in here and expect them to be
+/// dereferencable.
+#[derive(Debug)]
+#[deku_derive(DekuWrite)]
+struct YkSymFile {
+    sym_name: CString,
+    jitted_code_vaddr: usize,
+    jitted_code_size: usize,
+    src_path: std::ffi::CString,
+    num_lineinfos: c_int,
+    #[deku(count = "num_lineinfos")]
+    lineinfos: Vec<LineInfo>,
+}
+
+/// The gdb context for a trace.
+///
+/// This contains resources that need to be kept-alive for the trace to be safely debuggable.
+#[derive(Debug)]
+pub(crate) struct GdbCtx {
+    /// The file containing the "source code" that we show in gdb.
+    #[allow(dead_code)]
+    src_file: tempfile::NamedTempFile,
+    /// The entry into gdb's linked list.
+    ///
+    /// This is a ptr to a Rust `Box` that must be reconstituted to be freed.
+    #[allow(dead_code)]
+    jit_code_entry: *mut JitCodeEntry,
+    /// The serialised payload of the symbol file.
+    #[allow(dead_code)]
+    payload: Vec<u8>,
+}
+
+unsafe impl Send for GdbCtx {}
```

Comment:
These are new since you last saw this PR. This was caused by using `Box::into_raw()` (as we discussed earlier), which means be have to store a raw pointer instead of a Box.

I haven't been able to convince myself it's absolutely safe.

Reply:
*Ideally* we'd only Send/Sync the pointer, and not the other parts of the struct. A lazy way of doing this would be to store the `*mut` as a `usize` and then you wouldn't need the `unsafe impl`s!